Migrate simple and dominator based jump threading to OSSA

meg-gupta opened the pull request:
Migrate remaining SimplifyCFG transforms to OSSA.

meg-gupta committed:
We don't consider it profitable to jump thread when we could optimize arc in OSSA.
Copy propagation should be able to handle such cases without jump threading.

meg-gupta committed:
Fixes to SimplifyCFG::threadEdge for OSSA

meg-gupta committed:
Disable SimplifyCFG::tailDuplicateObjCMethodCallSuccessorBlocks in OSSA

Support for this needs to be added

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test source compatibility

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented:
@swift-ci Please Apple Silicon benchmark

meg-gupta commented:
@swift-ci test

meg-gupta committed:
Update tests

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented:
@swift-ci test source compatibility

meg-gupta committed:
Fix simplifyArguments for a few edges cases in OSSA

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta committed:
Remove unstable irgen diagnostic swift test, add a sil test instead

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci please test source compatibility

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented:
@swift-ci please apple silicon benchmark

meg-gupta commented:
REGRESSION                                 OLD         NEW         DELTA     RATIO    
KeyPathsSmallStruct                        2.396       9.488       +295.9%   **0.25x**
KeyPathNestedStructs                       1.103       3.014       +173.1%   **0.37x**
ParseInt.IntSmall.Decimal                  87.739      156.0       +77.8%    **0.56x**
KeyPathReadPerformance                     11.543      19.115      +65.6%    **0.60x (?)**
ParseInt.IntSmall.UncommonRadix            105.632     171.333     +62.2%    **0.62x**
ParseInt.UIntSmall.Binary                  230.3       335.286     +45.6%    **0.69x**
Set.isStrictSubset.Int.Empty               29.615      37.429      +26.4%    **0.79x (?)**
RemoveWhereQuadraticString                 182.5       229.0       +25.5%    **0.80x (?)**
Set.isStrictSubset.Int0                    32.727      40.548      +23.9%    **0.81x (?)**
Set.isSubset.Seq.Empty.Int                 32.727      40.545      +23.9%    **0.81x (?)**
Set.isSubset.Int.Empty                     34.292      42.103      +22.8%    **0.81x (?)**
Set.isSuperset.Seq.Empty.Int               34.292      42.103      +22.8%    **0.81x (?)**
Set.isStrictSubset.Box0                    34.286      42.095      +22.8%    **0.81x (?)**
Set.isDisjoint.Seq.Int.Empty               34.292      42.1        +22.8%    **0.81x (?)**
Set.isDisjoint.Seq.Empty.Int               37.409      45.231      +20.9%    **0.83x (?)**
Set.isSuperset.Seq.Int.Empty               39.0        46.789      +20.0%    **0.83x (?)**
Set.isDisjoint.Seq.Empty.Box               39.0        46.773      +19.9%    **0.83x (?)**
Set.isDisjoint.Int.Empty                   35.87       42.119      +17.4%    **0.85x (?)**
Set.isDisjoint.Empty.Int                   39.0        45.227      +16.0%    **0.86x (?)**
Set.isDisjoint.Box.Empty                   39.0        45.222      +16.0%    **0.86x (?)**
Set.isDisjoint.Empty.Box                   40.548      46.786      +15.4%    **0.87x**
Set.isDisjoint.Seq.Box.Empty               57.706      65.5        +13.5%    **0.88x (?)**
Set.isSubset.Empty.Int                     68.583      76.417      +11.4%    **0.90x (?)**
Set.isStrictSubset.Empty.Int               68.583      76.417      +11.4%    **0.90x (?)**
Set.isStrictSubset.Seq.Int.Empty           81.091      88.909      +9.6%     **0.91x (?)**
Set.isStrictSuperset.Seq.Empty.Int         81.091      88.9        +9.6%     **0.91x (?)**
Set.isSubset.Seq.Int.Empty                 81.091      88.9        +9.6%     **0.91x (?)**
Set.subtracting.Int.Empty                  17.462      19.024      +8.9%     **0.92x (?)**
Set.subtracting.Seq.Empty.Int              90.444      98.25       +8.6%     **0.92x (?)**
Set.subtracting.Seq.Int.Empty              95.9        103.824     +8.3%     **0.92x (?)**
Set.subtracting.Empty.Int                  19.324      20.886      +8.1%     **0.93x (?)**
ObjectiveCBridgeStringHash                 42.313      45.722      +8.1%     **0.93x (?)**
StringFromLongWholeSubstring               2.018       2.172       +7.6%     **0.93x (?)**

IMPROVEMENT                                OLD         NEW         DELTA     RATIO    
RemoveWhereFilterString                    193.417     158.067     -18.3%    **1.22x**
SIMDReduce.Int32x16.Initializer            13.408      11.07       -17.4%    **1.21x (?)**
LineSink.characters.complex                290.25      241.4       -16.8%    **1.20x**
LineSink.characters.alpha                  65.306      54.409      -16.7%    **1.20x**
String.replaceSubrange.Substring           7.784       6.69        -14.1%    **1.16x (?)**
ArrayAppendLazyMap                         710.4       613.947     -13.6%    **1.16x (?)**
RawBufferCopyBytes                         11.944      10.361      -13.3%    **1.15x (?)**
ArrayAppendGenericStructs                  473.529     412.941     -12.8%    **1.15x (?)**
StringMatch                                3940.0      3497.222    -11.2%    **1.13x (?)**
String.initRepeating.1AsciiChar.Count100   291.667     259.375     -11.1%    **1.12x (?)**
FindString.Loop1.Substring                 290.333     259.0       -10.8%    **1.12x (?)**
StringDistance.characters.ascii            2319.0      2089.0      -9.9%     **1.11x (?)**
Diffing.PangramToAlphabet                  687.0       619.333     -9.8%     **1.11x (?)**
Diffing.ReversedAlphabets                  117.9       107.0       -9.2%     **1.10x (?)**
Diffing.Disparate                          97.273      88.917      -8.6%     **1.09x (?)**
StringWordBuilderReservingCapacity         733.333     671.0       -8.5%     **1.09x (?)**
StringWordBuilder                          748.571     685.714     -8.4%     **1.09x (?)**
StringInterpolationManySmallSegments       3440.0      3153.846    -8.3%     **1.09x (?)**
AngryPhonebook.Strasse                     332.571     305.0       -8.3%     **1.09x**
StringHasPrefixAscii                       1153.75     1060.625    -8.1%     **1.09x (?)**
StringHasSuffixUnicode                     54916.667   50615.385   -7.8%     **1.08x (?)**
SubstringFromLongStringGeneric2            20.269      18.697      -7.8%     **1.08x (?)**
StringHasSuffixAscii                       1216.25     1122.5      -7.7%     **1.08x (?)**
LuhnAlgoEager                              85.706      79.5        -7.2%     **1.08x (?)**
RomanNumbers2                              310.857     288.571     -7.2%     **1.08x (?)**
String.replaceSubrange.String.Small        24.48       22.778      -7.0%     **1.07x (?)**
LuhnAlgoLazy                               85.7        80.043      -6.6%     **1.07x (?)**

meg-gupta commented:
TODO: Investigate regressions

atrick commented on lib/SILOptimizer/Transforms/SimplifyCFG.cpp:
```diff
@@ -231,7 +232,8 @@ bool SimplifyCFG::threadEdge(const ThreadInfo &ti) {
         assert(SEI->getDefaultBB() == ThreadedSuccessorBlock);
-        Builder.createBranch(SEI->getLoc(), ThreadedSuccessorBlock, SEI->getOperand());
+        Builder.createBranch(SEI->getLoc(), ThreadedSuccessorBlock,
+                             SEI->getOperand());
```
Since phis mean something in OSSA more than simply a branch argument. We should either change this implementation to RAUW without creating a phi. Or explicitly initialize a new phi.

atrick replied:
It looks like this commit fixed the places where we were accidentally creating new phis without properly creating the phi:
https://github.com/apple/swift/pull/62959/commits/9311e948a55aa548a38c15733d34620b949384c7

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented:
@swift-ci apple silicon benchmark

meg-gupta commented:
@swift-ci apple silicon benchmark

meg-gupta commented:
@swift-ci apple silicon benchmark

meg-gupta committed:
Add @always(__inline) to _parseInteger(UnsafeBufferPointer<UInt8>, Int)

Integer parsing seems to be very sensitive to inlining.
With simplifycfg, _parseInteger function gets bigger and ends up not being inlined
causing significant performance regression in the integer parsing benchmarks.

Add @always(__inline) to recover the performance.

meg-gupta commented:
@swift-ci benchmark

meg-gupta committed:
Add new flags to enable specific SimplifyCFG operations

meg-gupta committed:
Disable ArgumentSplitter for non trivial types in ossa

meg-gupta commented:
@swift-ci Please Apple Silicon benchmark

meg-gupta committed:
Avoid unnecessary block arguments in SimplifyCFG::threadEdge and SimplifyCFG::simplifyThreadedTerminators

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

atrick reviewed (approved):
Looks good. Congrats on finishing this big milestone pass!

meg-gupta committed:
Add requires to irgen codegen test

meg-gupta commented:
@swift-ci smoke test and merge
